pkg/iss/iss.go

```diff
@@ -605,6 +596,36 @@ func (iss *ISS) applyStableCheckpoint(stableCheckpoint *isspb.StableCheckpoint)
 	return eventsOut
 }
 
+func (iss *ISS) applyPushCheckpoint() (*events.EventList, error) {
+
+	// Schedule the next timer event.
+	// The Repeat function of the Timer is not used here,
+	// because this periodic event cannot be explicitly garbage-collected.
+	eventsOut := events.ListOf(events.TimerDelay(
+		timerModuleName,
+		[]*eventpb.Event{PushCheckpoint()},
+		t.TimeDuration(iss.config.CatchUpTimerPeriod),
+	))
```

Comment:
This potentially delays sending the stable checkpoint up to `CatchUpTimerPeriod`. I think delayed nodes should get assistance as soon as possible, since the more we wait the more delayed they become, and, under high load, by the time they receive the stable checkpoint they may be already too far behind again. 

Moreover, I don't think we can guarantee that the time it takes to finish an epoch is strictly less than `CatchUpTimerPeriod / RetainedEpochs`.

Reply:
The former implementation sent the StableCheckpoint message right away. We could have just added periodic re-transmission of that message. 

Again, under heavy load, it may be too late to try catching up after a few seconds. Namely, if the rest of the system keeps advancing for more than `RetainedEpochs` while waiting for `CatchUpTimerPeriod` before sending StableCheckpoint messages to the delayed node then the delayed node has no chance to catch up.

Reply:
Or even without re-transmission, the delayed node would just need to wait until the next stable checkpoint is received in time.

Reply:
Fair enough. I still think that even the independent catchup mechanism works in practice and I find it cognitively simpler. But it's true that sending the checkpoint right away helps make the situation you described less likely. Simply periodically re-transmitting the message with the checkpoint could get rather wasteful, should the checkpoint consist of more than just a simple digest. I thus implemented the direct catching up right after the stable checkpoint, but the re-transmission still happens using a periodic internal event upon which the node checks whom to send the checkpoint to.

Reply:
Let it be)